Sources/NIOHTTPCompression/HTTPCompression.swift

```diff
@@ -0,0 +1,148 @@
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import CNIOExtrasZlib
+import NIO
+
+public enum NIOHTTPCompressionSettings {
+
+    public struct CompressionAlgorithm: CustomStringConvertible {
+        enum Algorithm: String {
+            case gzip
+            case deflate
+        }
+        let algorithm: Algorithm
+        
+        /// return as String
+        public var description: String { return algorithm.rawValue }
+        
+        public static var gzip = CompressionAlgorithm(algorithm: .gzip)
+        public static var deflate = CompressionAlgorithm(algorithm: .deflate)
+    }
+        
+    public struct CompressionError: Error, CustomStringConvertible, Equatable {
+        enum ErrorType: String {
+            case uncompressedWritesPending
+            case noDataToWrite
+        }
+        let error: ErrorType
+        
+        /// return as String
+        public var description: String { return error.rawValue }
+        
+        public static var uncompressedWritesPending = CompressionError(error: .uncompressedWritesPending)
+        public static var noDataToWrite = CompressionError(error: .noDataToWrite)
+    }
+        
+    struct Compressor {
+        private var stream = z_stream()
+        var isActive = false
+
+        init() { }
+
+        /// Set up the encoder for compressing data according to a specific
+        /// algorithm.
+        mutating func initialize(encoding: CompressionAlgorithm) {
+            assert(!isActive)
+            isActive = true
+            // zlib docs say: The application must initialize zalloc, zfree and opaque before calling the init function.
+            stream.zalloc = nil
+            stream.zfree = nil
+            stream.opaque = nil
+
+            let windowBits: Int32
+            switch encoding.algorithm {
+            case .deflate:
+                windowBits = 15
+            case .gzip:
+                windowBits = 16 + 15
+            }
+
+            let rc = CNIOExtrasZlib_deflateInit2(&stream, Z_DEFAULT_COMPRESSION, Z_DEFLATED, windowBits, 8, Z_DEFAULT_STRATEGY)
+            precondition(rc == Z_OK, "Unexpected return from zlib init: \(rc)")
+        }
+
+        mutating func compress(inputBuffer: inout ByteBuffer, allocator: ByteBufferAllocator, finalise: Bool) -> ByteBuffer {
+            assert(isActive)
+            let flags = finalise ? Z_FINISH : Z_SYNC_FLUSH
+            guard inputBuffer.readableBytes > 0 || finalise == true else { return allocator.buffer(capacity: 0) }
+            // deflateBound() provides an upper limit on the number of bytes the input can
+            // compress to. We add 5 bytes to handle the fact that Z_SYNC_FLUSH will append
+            // an empty stored block that is 5 bytes long.
```

Comment:
The extra 5 bytes aren't added below. It's also it's not clear to me why adding 5 bytes would be okay here in light of the above comment; am I missing something?

Reply:
This was code moved across from HTTPResponseCompressor. I wanted to share code between the request and response compressors. I went to look for the original author of this and it was @weissi but his comment is copied to swift-nio-extras so the source of this is a little vague. I will look into zlib docs to see if I can get a better idea what this was for.

Reply:
See comment above